File: client/kotlin/src/app/kntrl/client/Session.kt
package app.kntrl.client

import app.kntrl.client.openapi.api.AuthorisationApi
import app.kntrl.client.openapi.api.SessionApi
import app.kntrl.client.openapi.api.TokenApi
import app.kntrl.client.openapi.infra.ApiClient
import app.kntrl.client.openapi.model.*
import app.kntrl.client.openapi.model.Session as SessionModel

class Session(
    private val client: ApiClient,
    private var tokens: Tokens?, private var newSessionReq: NewSessionReq?,
) : SessionModel() {
    val server = ServerSvc(this)
    @get:JvmName("user") val user = UserSvc(this)
    val rateLimiter = RateLimiterSvc(this)

    init {
        id = ""
        entry = newSessionReq?.entry ?: ""
        identifiedBy = emptyList()
        userId = ""
        authenticatedBy = emptyMap()
        unconfirmedAuths = emptyMap()
        nextFactors = emptyList()
        signedInAt = 0
        expiresAt = 0
        refreshedAt = 0
        newUser = false
        device = Device().type("").userAgent("").browser("").mobile(false)
        systemAccess = false
        expired = false
        authenticated = false
    }

    fun get(): SessionModel = handleErr(this) {
        val res = SessionApi(authenticatedOpenapiClient()).session
        update(null, res)
        res
    }


    fun authenticate(req: AuthenticateReq): AuthenticateRes = handleErr(this) {
        val res = if (newSessionReq != null) {
            SessionApi(authenticatedOpenapiClient()).newSession(newSessionReq!!.apply {
                factors = req.factors
                authReqs = req.authReqs
                dryRun = req.dryRun
            })
        } else {
            SessionApi(authenticatedOpenapiClient()).authenticate(req)
        }
        if (res.tokens?.access != null) newSessionReq = null
        update(res.tokens, res.session)
        res
    }

    fun confirm(receivedCodes: Map<String, Map<String, String>>): AuthenticateRes = confirm(null, receivedCodes)
    fun confirm(sessionId: String?, receivedCodes: Map<String, Map<String, String>>): AuthenticateRes = handleErr(this) {
        val res = SessionApi(authenticatedOpenapiClient()).confirmSessionAuths(
            ConfirmSessionAuthsReq().sessionId(sessionId).receivedCodes(receivedCodes))
        update(res.tokens, res.session)
        res
    }

    fun signOut() = signOut(null)
    fun signOut(sessionId: String?) = handleErr(this) {
        if (sessionId != null) {
            SessionApi(authenticatedOpenapiClient()).delSession(sessionId)
        } else {
            SessionApi(authenticatedOpenapiClient()).signOut()
        }
    }

    fun refreshAccessToken() = refreshAccessToken(null)
    fun refreshAccessToken(refreshToken: String?) = refreshAccessToken(refreshToken, null)
    fun refreshAccessToken(
        overrideRefreshToken: String?,
        errOnMissingToken: Exception?,
    ): RefreshTokenRes = handleErr {
        val refreshToken = overrideRefreshToken
            ?: tokens?.refresh
            ?: throw (errOnMissingToken ?: IllegalArgumentException("No refresh token"))

        val res = TokenApi(client).refreshToken(RefreshTokenReq().refreshToken(refreshToken))
        update(res.tokens, res.session)
        res
    }

    fun accessToken() = accessToken(1000)
    fun accessToken(refreshBeforeExpireMs: Long): String? {
        if (tokens?.refresh == null) return this.tokens?.access;

        if (tokens!!.accessTokenExpiresAt!! - System.currentTimeMillis() < refreshBeforeExpireMs) {
            this.refreshAccessToken()
        }
        return tokens?.access
    }

    fun newSession(entry: String) = Session(
        client, tokens,
        NewSessionReq().entry(entry).signIn(true),
    )

    fun authorize() = authorize(null)
    fun authorize(rateLimiter: RateLimiterReq?): AuthorizeRes = handleErr(this) {
        AuthorisationApi(authenticatedOpenapiClient())
            .authorize(AuthorizeReq().rateLimiter(rateLimiter))
    }

    fun allSession() = allSession(null, null)
    fun allSession(entry: String) = allSession(entry, null)
    fun allSession(entry: String?, userId: String?) = handleErr(this) {
        SessionApi(authenticatedOpenapiClient()).getAll(entry, userId)
    }

    fun authenticatedOpenapiClient() = if (this.tokens == null) {
        this.client
    } else {
        ApiClient(client.httpClient).apply {
            basePath = client.basePath
            setBearerToken(accessToken())
        }
    }

    private fun update(tokens: Tokens?, session: SessionModel?) = synchronized(this) {
        tokens?.also { this.tokens = tokens }
        session?.also {
            id = it.id
            entry = it.entry
            identifiedBy = it.identifiedBy
            userId = it.userId
            authenticatedBy = it.authenticatedBy
            unconfirmedAuths = it.unconfirmedAuths
            nextFactors = it.nextFactors
            signedInAt = it.signedInAt
            expiresAt = it.expiresAt
            refreshedAt = it.refreshedAt
            newUser = it.newUser
            device = it.device
            systemAccess = it.systemAccess
            expired = it.expired
            authenticated = it.authenticated
        }
    }

    class Factors : LinkedHashMap<String, String>() {
        fun factor(factor: String, auth: String): Factors {
            put(factor, auth)
            return this
        }
    }

    class AuthReqs : AuthenticateReq() {
        fun req(auth: String, req: AuthenticateReqAuthReqsValue?): AuthReqs {
            putAuthReqsItem(auth, req)
            return this
        }
        fun req(auth: String, factor: String, req: AuthenticateReqAuthReqsValue?): AuthReqs {
            putAuthReqsItem(auth, req)
            putFactorsItem(auth, factor)
            return this
        }
        fun enableFactor(factor: String): AuthReqs {
            putFactorsItem(factor, null)
            return this
        }
        fun dryRun(): AuthReqs {
            dryRun(true)
            return this
        }
    }
}

class AuthReq : AuthenticateReqAuthReqsValue()

File: client/typescript/src/session.ts
import {
    AuthenticateReq,
    AuthenticateRes,
    AuthorisationApi,
    AuthoriseReq,
    AuthoriseRes,
    Configuration,
    ConfirmSessionAuthsReq,
    Device,
    FindSessionsRes,
    LoginId,
    NewSessionReq,
    NextFactor,
    RefreshTokenRes,
    Session as SessionModel,
    SessionApi,
    TokenApi,
    Tokens,
    UnconfirmedAuth,
} from '../generated';
import { handleErr, merge } from './utils';
import { UserSvc } from './user';
import { ServerSvc } from './server';
import { AxiosRequestConfig } from 'axios';
import { RateLimiterSvc } from './rate-limiter';
import { Kntrl } from '../index';

export class Session implements SessionModel {
    private readonly api: SessionApi;

    constructor(
        private kntrl: Kntrl,
        private newSessionReq: Omit<NewSessionReq, keyof AuthenticateReq> | undefined,
        readonly tokens: Tokens | undefined = undefined,
        readonly id: string = '',
        readonly entry: string = newSessionReq?.entry ?? '',
        readonly identifiedBy: Array<LoginId> = [],
        readonly authenticatedBy: Record<string, string> = {},
        readonly userId: string | undefined = undefined,
        readonly unconfirmedAuths: Record<string, UnconfirmedAuth> = {},
        readonly nextFactors: Array<NextFactor> = [],
        readonly signedInAt: number = 0,
        readonly expiresAt: number = 0,
        readonly newUser: boolean = false,
        readonly authenticated: boolean = false,
        readonly device: Device = { userAgent: '' },
        readonly refreshedAt: number = 0,
        readonly systemAccess: boolean = false,
    ) {
        this.api = new SessionApi(kntrl._serverCfg());
        this.user = new UserSvc(this);
        this.server = new ServerSvc(this);
        this.rateLimiter = new RateLimiterSvc(this)
    }

    async get(): Promise<SessionModel> {
        const res = await handleErr(this.api.getSession(await this._authenticatedAxiosCfg()), this);
        merge(this, res);
        return res;
    }

    async authenticate(req: AuthenticateReq = {}): Promise<AuthenticateRes> {
        const res = await handleErr(async () => {
            if (this.newSessionReq) {
                return this.api.newSession({
                    ...this.newSessionReq,
                    ...req,
                }, await this._authenticatedAxiosCfg());
            } else {
                return this.api.authenticate(req, await this._authenticatedAxiosCfg());
            }
        }, this);
        if (res?.tokens?.access) this.newSessionReq = undefined
        this.update(res);
        return res;
    }

    async confirmAuth(req: ConfirmSessionAuthsReq): Promise<AuthenticateRes> {
        const res = await handleErr(this.api.confirmSessionAuths(req, await this._authenticatedAxiosCfg()), this);
        this.update(res);
        return res;
    }

    async signOut(): Promise<void> {
        await handleErr(this.api.signOut(await this._authenticatedAxiosCfg()), this);
    }

    async refreshAccessToken(refreshToken?: string, errOnMissingToken?: any): Promise<RefreshTokenRes> {
        refreshToken = refreshToken || this.tokens?.refresh;
        if (!refreshToken) throw errOnMissingToken || 'No refresh token';

        const res = await handleErr(
            new TokenApi(this._serverCfg()).refreshToken({ refreshToken }, await this.kntrl._axiosCfg())
        );
        this.update(res);
        return res;
    }

    newSession(req: { entry: string }): Session {
        return new Session(this.kntrl, { entry: req.entry, signIn: true }, this.tokens);
    }

    async allSessions(entry?: string, userId?: string): Promise<FindSessionsRes> {
        return (await handleErr(this.api.getAll(entry, userId, await this._authenticatedAxiosCfg()), this));
    }

    async authorise(req: AuthoriseReq = {}): Promise<AuthoriseRes> {
        return handleErr(new AuthorisationApi(this._serverCfg()).authorize(req, await this._authenticatedAxiosCfg()), this);
    }

    async accessToken(refreshBeforeExpireMs: number = 1000) {
        if (!this.tokens?.refresh) return this.tokens?.access;

        if (this.tokens.accessTokenExpiresAt! - Date.now() < refreshBeforeExpireMs) await this.refreshAccessToken();
        return this.tokens?.access;
    }

    async _authenticatedAxiosCfg(): Promise<AxiosRequestConfig> {
        if (!this.tokens) return this.kntrl._axiosCfg();
        return {
            ...this.kntrl._axiosCfg(),
            headers: {
                ...this.kntrl._axiosCfg().headers,
                Authorization: `Bearer ${await this.accessToken()}`,
            },
        }
    }
    _serverCfg(): Configuration { return this.kntrl._serverCfg() }

    user: UserSvc;
    server: ServerSvc;
    rateLimiter: RateLimiterSvc

    private update(res: { tokens?: Tokens, session?: SessionModel }) {
        merge(this, res.session);
        if (res.tokens) (<any>this).tokens = res.tokens;
    }
}